staging/src/k8s.io/kube-aggregator/pkg/apiserver/openapi_aggregator.go

```diff
@@ -173,98 +240,81 @@ func (s *openAPIAggregator) updateOpenAPISpec() error {
+func (s *openAPIAggregator) UpdateApiServiceSpec(apiServiceName string, spec *spec.Swagger, etag string) error {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
 
-func (r *inMemoryResponseWriter) String() string {
-	s := fmt.Sprintf("ResponseCode: %d", r.respCode)
-	if r.data != nil {
-		s += fmt.Sprintf(", Body: %s", string(r.data))
+	specInfo, existingService := s.openAPISpecs[apiServiceName]
+	if !existingService {
+		return fmt.Errorf("APIService %s does not exists", apiServiceName)
 	}
-	if r.header != nil {
-		s += fmt.Sprintf(", Header: %s", r.header)
-	}
-	return s
-}
 
-func (s *openAPIAggregator) handlerWithUser(handler http.Handler, info user.Info) http.Handler {
-	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
-		if ctx, ok := s.contextMapper.Get(req); ok {
-			s.contextMapper.Update(req, request.WithUser(ctx, info))
+	return s.tryUpdatingServiceSpecs(func() error {
+		s.openAPISpecs[apiServiceName] = &openAPISpecInfo{
+			apiService: specInfo.apiService,
+			spec:       spec,
+			handler:    specInfo.handler,
+			etag:       etag,
 		}
-		handler.ServeHTTP(w, req)
+		return s.updateOpenAPISpec()
 	})
 }
 
-// downloadOpenAPISpec downloads openAPI spec from /swagger.json endpoint of the given handler.
-func (s *openAPIAggregator) downloadOpenAPISpec(handler http.Handler) (*spec.Swagger, error) {
-	handler = s.handlerWithUser(handler, &user.DefaultInfo{Name: aggregatorUser})
-	handler = request.WithRequestContext(handler, s.contextMapper)
-	handler = http.TimeoutHandler(handler, specDownloadTimeout, "request timed out")
-
-	req, err := http.NewRequest("GET", "/swagger.json", nil)
-	if err != nil {
-		return nil, err
-	}
-	writer := newInMemoryResponseWriter()
-	handler.ServeHTTP(writer, req)
-
-	switch writer.respCode {
-	case http.StatusOK:
-		openApiSpec := &spec.Swagger{}
-		if err := json.Unmarshal(writer.data, openApiSpec); err != nil {
-			return nil, err
+// AddUpdateApiService adds or updates the api service. It is thread safe.
+func (s *openAPIAggregator) AddUpdateApiService(handler http.Handler, apiService *apiregistration.APIService) error {
+	s.mutex.Lock()
+	defer s.mutex.Unlock()
+
+	return s.tryUpdatingServiceSpecs(func() error {
+		if specInfo, existingService := s.openAPISpecs[apiService.Name]; existingService {
+			specInfo.handler = handler
+			specInfo.apiService = *apiService
+		} else {
+			s.openAPISpecs[apiService.Name] = &openAPISpecInfo{
+				apiService: *apiService,
+				handler:    handler,
+			}
 		}
-		return openApiSpec, nil
-	default:
-		return nil, fmt.Errorf("failed to retrive openAPI spec, http error: %s", writer.String())
-	}
+		return s.updateOpenAPISpec()
+	})
 }
 
-// loadApiServiceSpec loads OpenAPI spec for the given API Service and then updates aggregator's spec.
-func (s *openAPIAggregator) loadApiServiceSpec(handler http.Handler, apiService *apiregistration.APIService) error {
+// RemoveApiServiceSpec removes an api service from OpenAPI aggregation. It is thread safe.
```

Comment:
add: if it does not exist, no error is returned.

Reply:
fixed in next commit.